File: python/subband.py
import numpy as np

def scale(data, width, height):

    #Scaling HL_1 LH_1
    for j in range(0, int(height/2)):
        for i in range(int(width/2), width):
            data[j][i] = int(data[j][i] * 2) 
            data[j + int(height/2)][i - int(width/2)] = int(data[j + int(height/2)][i - int(width/2)] * 2)

    #Scaling HH_2
    for j in range(int(height/4), int(height/2)):
        for i in range(int(width/4), int(width/2)):
            data[j][i] *= 2 

    #Scaling HL_2 LH_2
    for j in range(0, int(height/4)):
        for i in range(int(width/4), int(width/2)):
            data[j][i] = int(data[j][i] * 4) 
            data[j + int(height/4)][i - int(width/4)] = int(data[j + int(height/4)][i - int(width/4)] * 4)

    #Scaling HH_3
    for j in range(int(height/8), int(height/4)):
        for i in range(int(width/8), int(width/4)):
            data[j][i] = int(data[j][i] * 4) 

    #LL_3 HL_3
    for j in range(0, int(height/8)):
        for i in range(0, int(width/4)):
            data[j][i] = int(data[j][i] * 8) 
    #LH_3
    for j in range(int(height/8), int(height/4)):
        for i in range(0, int(width/8)):
            data[j][i] = int(data[j][i] * 8) 
    return data

def rescale(data, width, height):

    #Scaling HL_1 LH_1
    for j in range(0, int(height/2)):
        for i in range(int(width/2), width):
            data[j][i] = int(data[j][i] / 2) 
            data[j + int(height/2)][i - int(width/2)] = int(data[j + int(height/2)][i - int(width/2)] / 2)

    #Scaling HH_2
    for j in range(int(height/4), int(height/2)):
        for i in range(int(width/4), int(width/2)):
            data[j][i] /= 2 

    #Scaling HL_2 LH_2
    for j in range(0, int(height/4)):
        for i in range(int(width/4), int(width/2)):
            data[j][i] = int(data[j][i] / 4) 
            data[j + int(height/4)][i - int(width/4)] = int(data[j + int(height/4)][i - int(width/4)] / 4)

    #Scaling HH_3
    for j in range(int(height/8), int(height/4)):
        for i in range(int(width/8), int(width/4)):
            data[j][i] = int(data[j][i] / 4) 

    #LL_3 HL_3
    for j in range(0, int(height/8)):
        for i in range(0, int(width/4)):
            data[j][i] = int(data[j][i] / 8) 
    #LH_3
    for j in range(int(height/8), int(height/4)):
        for i in range(0, int(width/8)):
            data[j][i] = int(data[j][i] / 8) 
    return data

File: core/src/subband.c
#include "subband.h"
#include "segment_header.h"

void subband_scale(SegmentHeader* headers, int16_t* data, size_t width, size_t height) {

    size_t width_2 = width>>1;
    size_t width_4 = width>>2;
    size_t width_8 = width>>3;

    size_t height_2 = height>>1;
    size_t height_4 = height>>2;
    size_t height_8 = height>>3;
    
    uint8_t HH_1 = 0;
    uint8_t HL_1 = 1;
    uint8_t LH_1 = 1;
    uint8_t HH_2 = 1;
    uint8_t HL_2 = 2;
    uint8_t LH_2 = 2;
    uint8_t HH_3 = 2;
    uint8_t HL_3 = 3;
    uint8_t LH_3 = 3;
    uint8_t LL_3 = 3;

    if(headers->header_4.custom_weights) {
        HH_1 = headers->header_4.custom_weight_HH_1;
        HL_1 = headers->header_4.custom_weight_HL_1;
        LH_1 = headers->header_4.custom_weight_LH_1;
        HH_2 = headers->header_4.custom_weight_HH_2;
        HL_2 = headers->header_4.custom_weight_HL_2;
        LH_2 = headers->header_4.custom_weight_LH_2;
        HH_3 = headers->header_4.custom_weight_HH_3;
        HL_3 = headers->header_4.custom_weight_HL_3;
        LH_3 = headers->header_4.custom_weight_LH_3;
        LL_3 = headers->header_4.custom_weight_LL_3;
    }

    //Scaling HH_1
    for(size_t j = height_2; j < height; ++j) {
        for(size_t i = width_2; i < width; ++i) {
            data[j * width + i] <<= HH_1;
        }
    }

    //HL_1
    for(size_t j = 0; j < height_2; ++j) {
        for(size_t i = width_2; i < width; ++i) {
            data[j * width + i] <<= HL_1;
        }
    }

    //LH_1
    for(size_t j = height_2; j < height; ++j) {
        for(size_t i = 0; i < width_2; ++i) {
           data[j * width + i] <<= LH_1; 
        }
    }
    //Scaling HH_2
    for(size_t j = height_4; j < height_2; ++j) {
        for(size_t i = width_4; i < width_2; ++i) {
            data[j * width + i] <<= HH_2;
        }
    }

    //HL_2
    for(size_t j = 0; j < height_4; ++j) {
        for(size_t i = width_4; i < width_2; ++i) {
            data[j * width + i] <<= HL_2;
        }
    }

    //LH_2
    for(size_t j = height_4; j < height_2; ++j) {
        for(size_t i = 0; i < width_4; ++i) {
           data[j * width + i] <<= LH_2; 
        }
    }

    //Scaling HH_3
    for(size_t j = height_8; j < height_4; ++j) {
        for(size_t i = width_8; i < width_4; ++i) {
            data[j * width + i] <<= HH_3;
        }
    }

    //HL_3
    for(size_t j = 0; j < height_8; ++j) {
        for(size_t i = width_8; i < width_4; ++i) {
            data[j * width + i] <<= HL_3;
        }
    }

    //LH_3
    for(size_t j = height_8; j < height_4; ++j) {
        for(size_t i = 0; i < width_8; ++i) {
           data[j * width + i] <<= LH_3; 
        }
    }

    //LL_3
    for(size_t j = 0; j < height_8; ++j) {
        for(size_t i = 0; i < width_8; ++i) {
            data[j * width + i] <<= LL_3;
        }
    }
}
